core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -212,14 +212,20 @@ private[spark] class TaskSchedulerImpl(
       val stage = taskSet.stageId
       val stageTaskSets =
         taskSetsByStageIdAndAttempt.getOrElseUpdate(stage, new HashMap[Int, TaskSetManager])
-      stageTaskSets(taskSet.stageAttemptId) = manager
-      val conflictingTaskSet = stageTaskSets.exists { case (_, ts) =>
-        ts.taskSet != taskSet && !ts.isZombie
-      }
-      if (conflictingTaskSet) {
-        throw new IllegalStateException(s"more than one active taskSet for stage $stage:" +
-          s" ${stageTaskSets.toSeq.map{_._2.taskSet.id}.mkString(",")}")
+
+      // Mark all the existing TaskSetManagers of this stage as zombie, as we are adding a new one.
+      // This is necessary to handle a corner case. Let's say a stage has 10 partitions and has 2
+      // TaskSetManagers: TSM1(zombie) and TSM2(active). TSM1 has a running task for partition 10
+      // and it completes. TSM2 finishes tasks for partition 1-19, and thinks he is still active
+      // because partition 10 is not completed yet. However, DAGScheduler gets task completion
+      // events for all the 10 partitions and thinks the stage is finished. If it's a shuffle stage
+      // and somehow it has missing map outputs, then DAGScheduler will resubmit it and create a
+      // TSM3 for it. As a stage can't have more than one active task set managers, we must mark
+      // TSM2 as zombie (it actually is).
```

Comment:
If TSM3 is created just after TSM2 finished partition 10, so, how does TSM3 know about the finished partition 10?

Reply:
It doesn't need to know, and Spark will just waste resource to run unnecessary tasks. The cluster will not crush.

That's why I said
> After this PR, #21131 becomes a pure optimization, to avoid launching unnecessary tasks. #22806 and #23871 are still valuable to improve this optimization.

Reply:
If TSM here is for a result stage, so when TSM2 finished partition 10 which commited output to HDFS, TSM3 would throw TaskCommitDeniedException due to launch task for partition 10. And I think this is what #22806 and #23871 try to fix.

Reply:
This PR focuses on fixing the potential occurrence of `java.lang.IllegalStateException: more than one active taskSet for stage`, which is described in https://issues.apache.org/jira/browse/SPARK-23433 .

https://issues.apache.org/jira/browse/SPARK-25250 remains unfixed and will be addressed in #22806 or #23871 .

Note that, SPARK-23433 can crush the cluster, even #22806 or #23871  can fix it as well, we need a simple fix and backport to 2.3/2.4.

SPARK-25250 is just a matter of wasting resource, we can keep the fix in master only.

Reply:
That's make sense and the `Update`.

Reply:
Yep it makes sense to fix the issue that this PR addresses alongwith the other PR's for SPARK-25250.